src/UpgradeRule/AddNamespaceRule.php

```diff
@@ -77,6 +79,43 @@ public function withRoot($root)
         return $this;
     }
 
+
+    public function beforeUpgrade(CollectionInterface $code, CodeChangeSet $changeset)
```

Comment:
Maybe make this the default implementation and leave `beforeUpgradeFile()` as the empty function to be overridden?